### ISSUE TITLE
Enhancement: [no-base-to-string] deprecate `restrict-template-expression` into this rule

### ISSUE THREAD

**Issue by abrahamguo** [typescript-eslint/typescript-eslint#9644]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-base-to-string/

### Description

The description of `no-base-to-string` from the docs is

> JavaScript will call toString() on an object when it is converted to a string, such as when + adding to a string or in ${} template literals. The default Object .toString() uses the format "[object Object]", which is often not what was intended. This rule reports on stringified values that aren't primitives and don't define a more useful .toString() method.

and the description for `restrict-template-expressions` is

> JavaScript automatically [converts an object to a string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#string_coercion) in a string context, such as when concatenating it with a string using + or embedding it in a template literal using ${}. The default toString() method of objects uses the format "[object Object]", which is often not what was intended. This rule reports on values used in a template literal string that aren't strings, optionally allowing other data types that provide useful stringification results.

It seems to me that, especially with the `allowArray` option I recently added to `restrict-template-expressions` and the upcoming `allow` option in #8556, `restrict-template-expressions` is becoming much more similar to `no-base-to-string` except that `restrict-template-expressions` only applies to interpolation while `no-base-to-string` covers interpolation, concatenation, and explicit `toString` calls.

From @JoshuaKGoldberg on [#8556](https://github.com/typescript-eslint/typescript-eslint/pull/8556/files#r1664176756):

> > This rule is getting more similar to no-base-to-string. Is that OK? Do we combine the rules?
>
> Good question... I've talked about it in the past with @bradzacher but don't recall where.
> 
> IIRC the main points against were that the rules technically do two different ideas: one restricts what you pass to template expressions; the other more broadly prevents odd .toString()s.
> 
> Personally, I'm not opposed to eventually unifying the two rules. It'd be nice to not have to manage two of them - especially now that, as you've noted, they've grown much more similar...

### Fail

```TypeScript
All existing examples from `no-base-to-string` and `restrict-template-expressions`.
```

### Pass

```TypeScript
All existing examples from `no-base-to-string` and `restrict-template-expressions`.
```

**Comment by JoshuaKGoldberg** [comment]
Yeah I've found it more and more confusing over the years when to use which rule. They're _technically_ different rules but _generally_ cover the same area of responsibility. I struggle to imagine a case where a project would strongly want one and not the other. 👍 from me with a request for more input from @typescript-eslint/triage-team.

**Comment by Josh-Cena** [comment]
`no-base-to-string` is a _correctness_ rule. There's no reason you would disable it, as the reports indicate actual bugs in your code (`"[object Object]"` showing up). (There are unfixable bugs with this rule too, like certain types lacking `toString()` declaration, but you can configure it away.)

`restrict-template-expression`, and its sister `restrict-plus-operands`, are _stylistic_ rules. While their union catches a subset of the mistakes of `no-base-to-string`, their main purpose is to _allowlist_ displayable types, while `no-base-to-string` creates a _blocklist_ of non-displayable types. By this nature, not everyone would be happy with these two rules (especially when we restrict common things like `never` and arrays by default, as well as many others that rely on implicit string coercion).

**Comment by bradzacher** [comment]
Yeah there's significant overlap but there's also distinction. Josh's description is apt.

I think a good summary - `no-base-to-string` is about code correctness, `restrict-template-expressions` is about presentational correctness.

It's often that I'd disable the latter but I'd rarely disable the former.

**Comment by JoshuaKGoldberg** [comment]
> `restrict-template-expression`, and its sister `restrict-plus-operands`, are _stylistic_ rules

Should we move them into the `stylisticTypeChecked` config then?

I think my root issue with the `no-base-to-string` rule is that both its name and functionality _appear_ to be roughly the union of `restrict-plus-operands` & `restrict-template-expressions`. 

> While their union catches a subset of the mistakes caught by `no-base-to-string`

Just confirming: it looks to me like the only part of `no-base-to-string` _not_ covered by the other two rules is explicit `.toString()` calls. Am I missing anything?

I'm wondering now, if we remove the plus operand and template expression checking from `no-base-to-string`, would the three rules no longer have any overlap? And if so, would that fix the common confusions around them?

**Comment by Josh-Cena** [comment]
> Just confirming: it looks to me like the only part of `no-base-to-string` _not_ covered by the other two rules is explicit `.toString()` calls. Am I missing anything?

It would, by principle, catch other things that do stringification, including https://github.com/typescript-eslint/typescript-eslint/issues/4314 and https://github.com/typescript-eslint/typescript-eslint/issues/3388.

> I'm wondering now, if we remove the plus operand and template expression checking from `no-base-to-string`, would the three rules no longer have any overlap? And if so, would that fix the common confusions around them?

No, because then you would not have a rule that only enforces "no `"[object Object]"` in string values". Again: the two stylistic rules are too strict to be considered correctness only (they have correctness components, sure, but they check beyond that). If we add options like `allowCustomStringifiers` to these two rules, then it may work, though in general I don't find it appetizing.

**Comment by JoshuaKGoldberg** [comment]
> > Just confirming: it looks to me like the only part of `no-base-to-string` _not_ covered by the other two rules is explicit `.toString()` calls. Am I missing anything?
>
> It would, by principle, catch other things that do stringification, including #4314 and #3388.

Ah, gotcha, in that case we can consider my proposal to be explicit `.toString()` calls _and similar, such as `String(...)`_.

> they check beyond that

Just for clarity, what do you see as the stylistic points beyond the correctness checking?

I'm having a hard time conceptualizing why `console.log("Value: " + [])` is different from `console.log(\`Value: ${[]}\`) _(and equivalent array/etc. values)_.

**Comment by Josh-Cena** [comment]
The point is that ``console.log(`Value: ${{ toString: () => "Hi" }}`)`` should be allowed by default, which is the case with `no-base-to-string`. Banning it and requiring an explicit `String()` or `toString()` is a stylistic preference.

**Comment by JoshuaKGoldberg** [comment]
😅 I'm having a hard time keeping track of the different areas of concerns across:

* Logical vs. stylistic
* `Object.prototype.toString` vs implicit string coercion
* Numeric vs. stringification

...given that this impacts at least:

* `no-base-to-string`
* `restrict-plus-operands`
* `restrict-template-expressions`

I also personally am focusing on other work around project services and Promises over this issue area.

It'd be useful if someone could summarize all the things that we would want to lint for, along with whether they're logical or stylistic, and which rule(s) catch them now.

**Comment by Josh-Cena** [comment]
This is a stylistic case:

```js
const obj = { toString: () => "foo" };
`${obj}`; // restrict-template-expressions
`${obj.toString()}`; // all happy
obj + ""; // restrict-plus-operands
obj.toString(); // all happy
```

This is a correctness case:

```js
const obj = {};
`${obj}`; // restrict-template-expressions + no-base-to-string
`${obj.toString()}`; // no-base-to-string
obj + ""; // restrict-plus-operands + no-base-to-string
obj.toString(); // no-base-to-string
```

This is a stylistic (arguably) case:

```js
const obj = 1;
`${obj}`; // restrict-template-expressions
`${obj.toString()}`; // all happy
obj + ""; // restrict-plus-operands
obj.toString(); // all happy
```

TLDR is that `no-base-to-string` catches a well-defined set of cases known to be wrong, while `restrict-*` defines a well-defined set of cases known to be right.

**Comment by JoshuaKGoldberg** [comment]
> TLDR is that `no-base-to-string` catches a well-defined set of cases known to be wrong, while `restrict-*` allows a well-defined set of cases known to be right.

This delineation I like. I also like the delineation from https://github.com/typescript-eslint/typescript-eslint/issues/9644#issuecomment-2306226577 as one area being logical, the other stylistic.

Proposal @typescript-eslint/triage-team: let's mark this as accepting PRs for our next breaking-changes-allowed major to:

* Solidify the general concatenation/stringification/template-strings rules into two rules
* Have one of those rules be logical, the other stylistic
* Write detailed docs and a blog post explaining the history of them

?

**Comment by JoshuaKGoldberg** [comment]
We chatted internally. There doesn't seem to be consensus on a best way to delineate these rules' areas of functionality. It's not really clearly split between "logical" and "stylistic". I'm out of ideas and pretty much fine having them remain as-is.

We're pretty certain that if something does happen, it won't be a rule deprecation. If anybody else has a clear proposal for how to do things, please do feel free to file a new issue. This one is getting long 😅. 

Thanks all for the discussion!
